predictive_tracking_V2.py

```diff
@@ -0,0 +1,321 @@
+"""
+Predictive Predator Tracker
+Author: Ben McEwen
+
+The class predator tracker takes mp4 file input (standard extract.py format). 
+The centroid of the animal is determined (find_center()) and a Kalman filter is used to smooth the
+movement and reduce noise introduced by occlusion (kalman_filter()). 
+A moving average filter and Kalman filter is used to find the average velocity and this is used to 
+calculate the future position of the animal at a given time (predict_location()).
+"""
+
+
+import cv2
+import numpy as np
+import math
+import matplotlib.pyplot as plt
+import time
+
+from kalman.main import main
+
+class Predator_tracker:
+    """Predator tracker class: - Finds centroid of animal
+                               - Uses kalman filter to smooth effects of occlusion
+                               - Determines direction vector of animal
+                               - Predicts future position of animal"""
+    Green = (0,255,0)
+    Red = (0,0,255)
+    Blue = (255,0,0)
+    
+    def __init__(self, filename):
+        # Initialisation
+        self.filename = filename
+
+
+    def find_center(self, frame):
+        gray_image = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+    
+        # convert the grayscale image to binary image
+        ret, thresh = cv2.threshold(gray_image,127,255,0)
+    
+        # calculate moments of binary image
+        M = cv2.moments(thresh)
+        
+        # calculate x,y coordinate of center
+        if M["m00"] != 0:
+            cX = int(M["m10"] / M["m00"])
+            cY = int(M["m01"] / M["m00"])
+        else:
+            cX, cY = 0, 0
+        
+        center = np.array([int(cX),int(cY)])
+        radius = 1
+    
+        return center, frame
+    
+    
+    
+    def find_direction(self, frame, center, previous_center):
+        # Draw vector representing animals direction
+        length = 1000 
+        
+        # Calculate angle of line using current and previous point
+        diff = center - previous_center
+        if ((diff[0] != 0)):
+            angle = math.degrees(math.atan(diff[1]/diff[0]))
+            print('Angle: ', angle)
+        else:
+            angle = 0
+            print('Division by zero, angle set to 0')
+             
+        # Project point in front of animal 
+        diff[0] = int(length*(math.cos(math.radians(angle))))
+        diff[1] = int(length*(math.sin(math.radians(angle))))    
+        point = center - diff
+        
+        # Plot line between current center and projected point
+        cv2.line(frame, tuple(point), tuple(center), (0, 0, 0), 2)
+        
+        return angle
+        
+        
+        
+    def plot_results(self, real_x, real_y, estimated_x, estimated_y):
+        # plot Animals real location and estimated location
+        Real, = plt.plot(real_x[3:], real_y[3:], 'b-', label='Line 1')
+        Estimate, = plt.plot(estimated_x[3:], estimated_y[3:], 'r-', label='Line 2')
```

Comment:
variable names should be in all lower case with underscores between words, so `real` and `estimate` in this case.